pkg/chains/storage/grafeas/grafeas.go

```diff
@@ -0,0 +1,384 @@
+	}
+
+	// for in-toto
+	return attestationpb.GenericSignedAttestation_CONTENT_TYPE_UNSPECIFIED
+}
+
+// retrieve all occurrences created under a taskrun by filtering resource URI
+func (b *Backend) getOccurrences(ctx context.Context) ([]*pb.Occurrence, error) {
+	result := []*pb.Occurrence{}
+
+	// step 1: get all resource URIs created under the taskrun
+	uriFilters := []string{}
+	uriFilters = append(uriFilters, b.retrieveAllOCIURIs()...)
+	uriFilters = append(uriFilters, b.getTaskRunURI())
+
+	// step 2: find all occurrences by using ListOccurrences filters, and put them in result
+	for _, uriFilter := range uriFilters {
+		occs, err := b.findOccurrencesForCriteria(ctx, b.getProjectPath(), uriFilter)
+		if err != nil {
+			return result, err
+		}
+		result = append(result, occs...)
+	}
+
+	return result, nil
+}
+
+// find all occurrences based on a number of criteria
+// - current criteria we use are just project name and resource uri
+// - we can add more criteria later if we want i.e. occurrence Kind, severity and PageSize etc.
+func (b *Backend) findOccurrencesForCriteria(ctx context.Context, projectPath string, resourceURI string) ([]*pb.Occurrence, error) {
+	occurences, err := b.client.ListOccurrences(ctx,
+		&pb.ListOccurrencesRequest{
+			Parent: projectPath,
+			Filter: fmt.Sprintf("resourceUrl=%q", resourceURI),
+		},
+	)
+	if err != nil {
+		return nil, err
+	}
+	return occurences.GetOccurrences(), nil
+}
+
+// get resource uri for a taskrun in the format of namespace-scoped resource uri
+// `/apis/GROUP/VERSION/namespaces/NAMESPACE/RESOURCETYPE/NAME``
+// see more details here https://kubernetes.io/docs/reference/using-api/api-concepts/#resource-uris
+func (b *Backend) getTaskRunURI() string {
+	return fmt.Sprintf("/apis/%s/namespaces/%s/%s/%s@%s",
+		b.tr.GetGroupVersionKind().GroupVersion().String(),
+		b.tr.GetNamespace(),
+		b.tr.GetGroupVersionKind().Kind,
+		b.tr.GetName(),
+		string(b.tr.UID),
+	)
+}
+
+// get resource uri for an oci image in the format of `IMAGE_URL@IMAGE_DIGEST`
+func (b *Backend) getOCIURI(opts config.StorageOpts) string {
+	imgs := b.retrieveAllOCIURIs()
+	for _, img := range imgs {
+		// get digest part of the image representation
+		digest := strings.Split(img, "sha256:")[1]
+
+		// for oci image, the key in StorageOpts will be the first 12 chars of digest
+		// so we want to compare
+		digestKey := digest[:12]
+		if digestKey == opts.Key {
+			return img
+		}
+	}
+	return ""
+}
+
+// get the uri of all images for a specific taskrun in the format of `IMAGE_URL@IMAGE_DIGEST`
+func (b *Backend) retrieveAllOCIURIs() []string {
+	result := []string{}
+	images := artifacts.ExtractOCIImagesFromResults(b.tr, b.logger)
+
+	for _, image := range images {
+		ref := image.(name.Digest)
+		result = append(result, ref.Name())
+	}
+
+	return result
+}
+
+func checkTrustedHost(server string) error {
+	for _, t := range trustedHosts {
+		if strings.HasSuffix(server, t) {
```

Comment:
This won't work, since you could construct a URL like `example.com/googleapis.com`. Use https://pkg.go.dev/net/url#URL.Hostname instead (and for now, don't worry about supporting wildcards).